### ISSUE TITLE
`vec_order()` on array inputs

### ISSUE THREAD

**Issue by DavisVaughan** [r-lib/vctrs#306]
For array-like inputs with >2D, the following breaks the invariant of:

```
vec_size(vec_order(x)) == vec_size(x)
```

This happens because `order_proxy()` calls `order()` on the proxy object (which is just the array), but `order()` flattens its input.

``` r
library(vctrs)

vec_order(data.frame(x = 1:3, y = 1:3))
#> [1] 1 2 3

vec_order(matrix(1:6, ncol = 2))
#> [1] 1 2 3 4 5 6

vec_order(array(1:8, c(2, 2, 2)))
#> [1] 1 2 3 4 5 6 7 8
```

This affects a few other things as well, like `vctrs:::xtfrm.vctrs_vctr` and `vec_sort()`

**Comment by DavisVaughan** [comment]
Once you have the array in a 2D form, I think it is easy to see how it should be ordered. But it's a bit tough to think about how this can be generalized to >2 dimensions. I think you have to make a choice along the way about what axis varies first. I'm not sure which is more R like.

``` r

# how to order this in a vctrs way?
# (altered to show the difference between the two methods)
x <- array(1:8, c(2, 2, 2))
x[2] <- 1
x[3] <- 5
x
#> , , 1
#> 
#>      [,1] [,2]
#> [1,]    1    5
#> [2,]    1    4
#> 
#> , , 2
#> 
#>      [,1] [,2]
#> [1,]    5    7
#> [2,]    6    8

# could flatten to 2D like this
# where the last dimension varies first
x_last_first <- cbind(
  x[,1,1], 
  x[,1,2], 
  x[,2,1], 
  x[,2,2]
)

x_last_first
#>      [,1] [,2] [,3] [,4]
#> [1,]    1    5    5    7
#> [2,]    1    6    4    8

# could flatten to 2D like this
# where the first dimension (besides rows)
# varies first
x_front_first <- cbind(
  x[,1,1], 
  x[,2,1], 
  x[,1,2], 
  x[,2,2]
)

x_front_first
#>      [,1] [,2] [,3] [,4]
#> [1,]    1    5    5    7
#> [2,]    1    4    6    8

# once you have them as 2D, ordering would look like
order(
  x_last_first[,1],
  x_last_first[,2],
  x_last_first[,3],
  x_last_first[,4]
)
#> [1] 1 2

order(
  x_front_first[,1],
  x_front_first[,2],
  x_front_first[,3],
  x_front_first[,4]
)
#> [1] 2 1
```

**Comment by hadley** [comment]
Are you sure arrays are orderable? It seems natural to treat them like lists to me.

**Comment by DavisVaughan** [comment]
Maybe they aren't orderable in this way that I am trying to force them to be. With xtensor/numpy you can sort arrays on an axis, but it essentially treats each "submatrix" separately rather than shoving it into a 2D structure like I did above:

``` r
library(rray)

x <- matrix(4:1, ncol = 2)
x
#>      [,1] [,2]
#> [1,]    4    2
#> [2,]    3    1

# sorts rows
rray_sort(x, axis = 1L)
#>      [,1] [,2]
#> [1,]    3    1
#> [2,]    4    2


y <- array(c(4:1, 1:4), c(2, 2, 2))
y
#> , , 1
#> 
#>      [,1] [,2]
#> [1,]    4    2
#> [2,]    3    1
#> 
#> , , 2
#> 
#>      [,1] [,2]
#> [1,]    1    3
#> [2,]    2    4

# but it does so within each submatrix separately
rray_sort(y, axis = 1L)
#> , , 1
#> 
#>      [,1] [,2]
#> [1,]    3    1
#> [2,]    4    2
#> 
#> , , 2
#> 
#>      [,1] [,2]
#> [1,]    1    3
#> [2,]    2    4
```

**Comment by lionel-** [comment]
That has been fixed by #365, which converts arrays to data frames to make them comparable. It uses `base::as.data.frame()` which flattens the arrays by varying earlier dimensions first:

```r
#> x
#> , , 1
#> 
#>      [,1] [,2]
#> [1,]    1    5
#> [2,]    1    4
#> 
#> , , 2
#> 
#>      [,1] [,2]
#> [1,]    5    7
#> [2,]    6    8

as.data.frame(x)
#>   V1 V2 V3 V4
#> 1  1  5  5  7
#> 2  1  4  6  8
```
